src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContextSerializer.cs

```diff
@@ -25,25 +24,97 @@ private DesigntimeLicenseContextSerializer()
         /// </summary>
         public static void Serialize(Stream o, string cryptoKey, DesigntimeLicenseContext context)
         {
-            IFormatter formatter = new BinaryFormatter();
+            if (LocalAppContextSwitches.BinaryFormatterEnabled)
+            {
+                IFormatter formatter = new BinaryFormatter();
 #pragma warning disable SYSLIB0011 // Issue https://github.com/dotnet/runtime/issues/39293 tracks finding an alternative to BinaryFormatter
-            formatter.Serialize(o, new object[] { cryptoKey, context._savedLicenseKeys });
+                formatter.Serialize(o, new object[] { cryptoKey, context._savedLicenseKeys });
 #pragma warning restore SYSLIB0011
+            }
+            else
+            {
+                using (BinaryWriter writer = new BinaryWriter(o, encoding: Text.Encoding.UTF8, leaveOpen: true))
+                {
+                    writer.Write(-1); // flag to identify BinaryWriter
+                    writer.Write(cryptoKey);
+                    writer.Write(context._savedLicenseKeys.Count);
+                    foreach (DictionaryEntry keyAndValue in context._savedLicenseKeys)
+                    {
+                        writer.Write(keyAndValue.Key.ToString());
+                        writer.Write(keyAndValue.Value.ToString());
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// During deserialization, the stream passed in may be binary formatted or may have used binary writer. This is quick test to discern between them.
+        /// </summary>
+        internal static bool StreamIsBinaryFormatted(Stream stream)
+        {
+            if (!stream.CanSeek)
+            {
+                throw new Exception("Expected a seek-able stream");
+            }
+            bool ret = true;
+
+            // For binary formatter, the first byte is the SerializationHeaderRecord and has a value 0
+            int firstByte = stream.ReadByte();
+            if (firstByte != 0)
+            {
+                ret = false;
+            }
+
+            stream.Seek(-1, SeekOrigin.Current);
+            return ret;
         }
 
         internal static void Deserialize(Stream o, string cryptoKey, RuntimeLicenseContext context)
         {
+            if (StreamIsBinaryFormatted(o))
+            {
+                try
```

Comment:
consider extracting the BF codepaths out into their own method, in the places you switch between the formats? it might be more readable, also it possibly avoids the BinaryFormatter type being loaded at all

Reply:
I see the logic of this, but I assume we need some sort of attribute to tell the linker when to trim/(not trim)? I need to go through the linker docs to find it :) I moved the BF codepaths to their own methods though.